src/group_theory/free_product.lean

```diff
@@ -147,4 +168,157 @@ instance : group (free_product G) :=
 
 end group
 
+namespace word
+
+/-- The empty reduced word. -/
+def empty : word M := { to_list := [], ne_one := λ _, false.elim, chain_ne := list.chain'_nil }
+
+instance : inhabited (word M) := ⟨empty⟩
+
+/-- A reduced word determines an element of the free product, given by multiplication. -/
+def prod (w : word M) : free_product M :=
+list.prod (w.to_list.map $ λ l, of l.snd)
+
+@[simp] lemma prod_nil : prod (empty : word M) = 1 := rfl
+
+/-- `fst_idx w` is `some i` if the first letter of `w` is `⟨i, m⟩` with `m : M i`. If `w` is empty
+then it's `none`. -/
+def fst_idx (w : word M) : option ι := w.to_list.head'.map sigma.fst
+
+lemma fst_idx_ne_iff {w : word M} {i} :
+  fst_idx w ≠ some i ↔ ∀ l ∈ w.to_list.head', i ≠ sigma.fst l :=
+not_iff_not.mp $ by simp [fst_idx]
+
+variable (M)
+
+/-- Given an index `i : ι`, `pair M i` is the type of pairs `(head, tail)` where `head : M i` and
+`tail : word M`, subject to the constraint that first letter of `tail` can't be `⟨i, m⟩`.
+By prepending `head` to `tail`, one obtains a new word. We'll show that any word can be uniquely
+obtained in this way. -/
+@[ext] structure pair (i : ι) :=
+(head : M i)
+(tail : word M)
+(fst_idx_ne : fst_idx tail ≠ some i)
+
+instance (i : ι) : inhabited (pair M i) := ⟨⟨1, empty, by tauto⟩⟩
+
+variable {M}
+
+variables [∀ i, decidable_eq (M i)]
+
+/-- Prepend `m : M i` to `w` assuming `not_head i w`. Does nothing if `m = 1`. -/
```

Comment:
Does the docstring really correspond to the function?

Reply:
Good catch, I changed the function and forgot to update the docstring.